README
NAME
    Marketplace::Rakuten - Interface to
    http://webservice.rakuten.de/documentation

VERSION
    Version 0.03

CATEGORIES
    The list of categories for the German marketplace can be downloaded from
    the http://api.rakuten.de/categories/csv/download.

    When adding or editing a product, you can pass the `rakuten_category_id'
    to `add_product' or `edit_product'

SYNOPSIS
        use Marketplace::Rakuten;
        my $rakuten = Marketplace::Rakuten->new(key => 'xxxxxxx',
                                                endpoint => 'http://webservice.rakuten.de/merchants/'
                                                );
        my $res = $rakuten->get_key_info;

ACCESSORS
  key
    The API key (required).

  endpoint
    The URL of the endpoint. Default to
    http://webservice.rakuten.de/merchants/

    If you need a specific api version, it could be something like
    http://webservice.rakuten.de/v2.05/merchants

  ua
    The user-agent. Built lazily.

METHODS
  api_call($method, $call, $data);
    Generic method to do any call. The first argument is the HTTP request
    method (get or post). The second argument is the path of the api, e.g.
    (`misc/getKeyInfo'). The third is the structure to send (an hashref).

    Return a Marketplace::Rakuten::Response object.

    The key is injected into the data hashref in any case.

  API CALLS
    Miscellaneous
    get_key_info

    Product creation
    add_product(\%data)
        http://webservice.rakuten.de/documentation/method/add_product

        Mandatory params: `name', `price', `description'

        Recommended: `product_art_no' (the sku)

        Return hashref: `product_id'

    add_product_image(\%data)
        http://webservice.rakuten.de/documentation/method/add_product_image

        Mandatory params: `product_art_no' (the sku) or `product_id'
        (rakuten id).

        Return hashref: `image_id'

    add_product_variant(\%data)
        http://webservice.rakuten.de/documentation/method/add_product_varian
        t

        Mandatory params: `product_art_no' (the sku) or `product_id',
        `price', `color', `label'

        Return hashref: `variant_id'

    add_product_variant_definition
        http://webservice.rakuten.de/documentation/method/add_product_varian
        t

        Mandatory params: `product_art_no' (the sku) or `product_id'

        Recommended: `variant_1' Size `variant_2' Color

        You need to call this method before trying to upload variants. The
        data returned are however just a boolean (kind of) success.

    add_product_multi_variant
        http://webservice.rakuten.de/documentation/method/add_product_multi_
        variant

        Mandatory params: `product_art_no' (the sku) or `product_id',
        `variation1_type', `variation1_value', `price'

        Recommended: `variant_art_no'

        Return hashref: `variant_id'

    add_product_link
        http://webservice.rakuten.de/documentation/method/add_product_link

        Mandatory params: `product_art_no' (the sku) or `product_id', `name'
        (100 chars), `url' (255 chars)

        Return hashref: `link_id'

    add_product_attribute
        http://webservice.rakuten.de/documentation/method/add_product_attrib
        ute

        Mandatory params: `product_art_no' (the sku) or `product_id',
        `title' (50 chars), `value'

        Return hashref: `attribute_id'

    Product editing
    The following methods require and id passed (sku or id) and work the
    same as for product adding, but no argument is mandatory.

    They all return just a structure with a boolean success (-1 is failure,
    1 is success).

    edit_product(\%data)
        Requires `product_id' or `product_art_no'

    edit_product_variant(\%data);
        Requires `variant_id' or `variant_art_no'

    edit_product_variant_definition(\%data)
        Requires `product_id' or `product_art_no'

    edit_product_multi_variant(\%data);
        Requires `variant_id' or `variant_art_no'

    edit_product_attribute
        Requires `attribute_id', `title', `value'

    Product deletion
    The following methods require just an id or sku. Return a boolean
    success.

    delete_product(\%data)
        Requires `product_id' or `product_art_no'

    delete_product_variant(\%data);
        Requires `variant_id' or `variant_art_no'

    delete_product_image(\%data)
        Requires `image_id' (returned by add_product_image).

    delete_product_link(\%data);
        Requires `link_id' (returned by add_product_link).

    delete_product_attribute
        Requires `attribute_id' (returned by add_product_attribute).

    Orders
    get_orders(\%params)
        No argument is required, but you probably want to pass something
        like

         { status => pending }

        See http://webservice.rakuten.de/documentation/method/get_orders for
        the full list of options.

        List of statuses:

        pending
        editable
            This is what you want to get for importing.

        shipped
        payout
            Order is paid. Unclear when the status switches to this one.

        cancelled

        The response is paginated and from here you get only the raw data.

        Use get_parsed_orders to get the full list of objects.

    get_parsed_orders(\%params)
        The hashref with the parameters is the same of `get_orders' (which
        get called). This method takes care of the pagination and return a
        list of Marketplace::Rakuten::Order objects. You can access the raw
        structures with $object->order.

    get_pending_orders
        Shortcut for $self->get_parsed_orders({ status => 'pending' });

    get_editable_orders
        Shortcut for $self->get_parsed_orders({ status => 'editable' });

    set_order_shipped(\%params)
        Required parameter: order_no (the rakuten's order number).

        Accepted parameters:

        dhl (boolean, true if Rakuten-DHL-Rahmenvertrag is used)
        carrier
        tracking_number
        tracking_url

    set_order_cancelled(\%params)
        Required paramater: `order_no' (rakuten's order number)

        Optional: `comment'

    set_order_returned(\%params)
        Required paramater: `order_no' (rakuten's order number) and `type'
        (`fully' or `partly'):

        http://webservice.rakuten.de/documentation/method/set_order_returned

    Category management
    add_shop_category(\%params)
        Required parameter: `name'

        http://webservice.rakuten.de/documentation/method/add_shop_category

        Returned: `shop_category_id'

    edit_shop_category(\%params)
        Required parameter `shop_category_id' or `external_shop_category_id'

        http://webservice.rakuten.de/documentation/method/edit_shop_category

    get_shop_categories
        No mandatory parameter.

        http://webservice.rakuten.de/documentation/method/get_shop_categorie
        s

    delete_shop_category
        Required parameter `shop_category_id' or `external_shop_category_id'

        http://webservice.rakuten.de/documentation/method/delete_shop_catego
        ry

    add_product_to_shop_category
        Required parameters: `shop_category_id' or
        `external_shop_category_id', `product_id' or `product_art_no'
        (Rakuten's id or our sku).

        http://webservice.rakuten.de/documentation/method/add_product_to_sho
        p_category

AUTHOR
    Marco Pessotto, `<melmothx at gmail.com>'

BUGS
    Please report any bugs or feature requests to `bug-marketplace-rakuten
    at rt.cpan.org', or through the web interface at
    http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Marketplace-Rakuten. I
    will be notified, and then you'll automatically be notified of progress
    on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Marketplace::Rakuten

    You can also look for information at:

    * RT: CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Marketplace-Rakuten

    * AnnoCPAN: Annotated CPAN documentation
        http://annocpan.org/dist/Marketplace-Rakuten

    * CPAN Ratings
        http://cpanratings.perl.org/d/Marketplace-Rakuten

    * Search CPAN
        http://search.cpan.org/dist/Marketplace-Rakuten/

ACKNOWLEDGEMENTS
LICENSE AND COPYRIGHT
    Copyright 2015 Marco Pessotto.

    This program is free software; you can redistribute it and/or modify it
    under the terms of the the Artistic License (2.0). You may obtain a copy
    of the full license at:

    http://www.perlfoundation.org/artistic_license_2_0

    Any use, modification, and distribution of the Standard or Modified
    Versions is governed by this Artistic License. By using, modifying or
    distributing the Package, you accept this license. Do not use, modify,
    or distribute the Package, if you do not accept this license.

    If your Modified Version has been derived from a Modified Version made
    by someone other than you, you are nevertheless required to ensure that
    your Modified Version complies with the requirements of this license.

    This license does not grant you the right to use any trademark, service
    mark, tradename, or logo of the Copyright Holder.

    This license includes the non-exclusive, worldwide, free-of-charge
    patent license to make, have made, use, offer to sell, sell, import and
    otherwise transfer the Package with respect to any patent claims
    licensable by the Copyright Holder that are necessarily infringed by the
    Package. If you institute patent litigation (including a cross-claim or
    counterclaim) against any party alleging that the Package constitutes
    direct or contributory patent infringement, then this Artistic License
    to you shall terminate on the date that such litigation is filed.

    Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
    AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
    THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
    PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
    YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
    CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
    EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Marketplace-Rakuten

0.03 Mon Jul 20 17:16:35 2015 CEST

     [DOCUMENTATION]

     * Add missing documentation for complete POD coverage (Marco Pessotto).

0.02 Mon Jul 20 08:29:41 2015 CEST

     [BUG FIXES]

     * Change prerequisite on HTTP::Tiny to first version providing
       "www_form_urlencode" method (Stefan Hornburg)

0.01 Fri Jul 17 16:35:45 2015 CEST

     * Initial release.


Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Marketplace::Rakuten',
    AUTHOR           => q{Marco Pessotto <melmothx@gmail.com>},
    VERSION_FROM     => 'lib/Marketplace/Rakuten.pm',
    ABSTRACT_FROM    => 'lib/Marketplace/Rakuten.pm',
    LICENSE          => 'Artistic_2_0',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'Moo' => 0,
        'namespace::clean' => 0,
        'MooX::Types::MooseLike' => 0,
        'HTTP::Tiny' => 0,
        'XML::LibXML::Simple' => 0,
        'DateTime' => 0,
        'DateTime::Format::ISO8601' => 0,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Marketplace-Rakuten-*' },
    META_MERGE => {
        resources => {
            repository  => 'https://github.com/interchange/Marketplace-Rakuten.git',
            bugtracker => 'https://github.com/interchange/Marketplace-Rakuten.git/issues',
            IRC => 'irc://irc.freenode.net/#interchange',
        },
    },
);

# Local Variables:
# cperl-indent-parens-as-block: 1
# End:


lib/Marketplace/Rakuten/Order.pm
package Marketplace::Rakuten::Order;

use strict;
use warnings;
use Data::Dumper;
use DateTime;
use DateTime::Format::ISO8601;

use Moo;
use MooX::Types::MooseLike::Base qw(Str HashRef Int Object);
use Marketplace::Rakuten::Order::Address;
use Marketplace::Rakuten::Order::Item;
use Marketplace::Rakuten::Utils;

use namespace::clean;

=head1 NAME

=encoding utf8

Marketplace::Rakuten::Order

=head1 DESCRIPTION

Class to handle the xml structures returned by
L<http://webservice.rakuten.de/documentation/method/get_orders>

The aim is to have a consistent interface with
L<Amazon::MWS::XML::Order> so importing the orders can happens almost
transparently.

=cut

=head1 ACCESSORS/METHODS

=head2 order

The raw structure got from the XML parsing

=head2 shop_type

Always returns C<rakuten>

=cut

has order => (is => 'ro', isa => HashRef, required => 1);

sub shop_type {
    return 'rakuten';
}

=head2 order_number

read-write accessor for the (shop) order number so you can set this
while importing it.

=cut

has order_number => (is => 'rw', isa => Str);

=head2 payment_status

read-write accessor for the payment status, so the shop can set it
while importing it.

=cut

has payment_status => (is => 'rw', isa => Str);

=head2 can_be_imported

=head2 order_status

Unclear (for now) what to do here. List of statuses:
 	
=over 4

=item pending

 Bestellung ist neu eingegangen

=item editable

Bestellung ist zur Bearbeitung freigegeben

=item shipped

Bestellung ist versendet

=item payout

Bestellung ist ausbezahlt

=item cancelled

Bestellung ist storniert

=back 

=cut

sub can_be_imported {
    my $self = shift;
    my %map = (
               pending => 1,
               editable => 1,
               payout => 1,
               cancelled => 0,
              );
    return $map{$self->order_status} || 0;
}

sub order_status {
    return shift->order->{status};
}

=head2 remote_shop_order_id

The Rakuten order id.

=cut

sub remote_shop_order_id {
    return shift->order->{order_no};
}

has shipping_address => (is => 'lazy');

sub _build_shipping_address {
    my $self = shift;
    my $address = $self->order->{delivery_address};
    my $billing = $self->order->{client};
    my %args;
    if ($address) {
        %args = %$address;
        # populate the object with billing data as well
        if ($billing) {
            foreach my $k (qw/client_id email phone/) {
                $args{$k} = $billing->{$k};
            }
        }
        return $self->_address_building_routine(%args);
    }
    return;
}

has billing_address => (is => 'lazy');

sub _build_billing_address {
    my $self = shift;
    my $billing = $self->order->{client};
    if ($billing) {
        return $self->_address_building_routine(%$billing);
    }
    return undef;
}

has items_ref => (is => 'lazy');

sub _address_building_routine {
    my ($self, %params) = @_;
    my %args = %params;
    Marketplace::Rakuten::Utils::turn_empty_hashrefs_into_empty_strings(\%args);
    return Marketplace::Rakuten::Order::Address->new(%args);
}


sub _build_items_ref {
    my ($self) = @_;
    my @items;
    my $order_num = $self->remote_shop_order_id;
    if ($self->order->{items}) {
        if (my $item = $self->order->{items}->{item}) {
            if (ref($item) eq 'HASH') {
                Marketplace::Rakuten::Utils::turn_empty_hashrefs_into_empty_strings($item);
                push @items,
                  Marketplace::Rakuten::Order::Item
                    ->new(struct => $item,
                          order_number => $order_num,
                         );
            }
            elsif (ref($item) eq 'ARRAY') {
                foreach my $i (@$item) {
                    Marketplace::Rakuten::Utils::turn_empty_hashrefs_into_empty_strings($i);
                    push @items,
                      Marketplace::Rakuten::Order::Item
                        ->new(struct => $i,
                              order_number => $order_num,
                             );
                }
            }
            else {
                die "Unexpected orderline" . Dumper($item);
            }
        }
    }
    return \@items;
}

sub items {
    return @{ shift->items_ref };
}

has number_of_items => (is => 'lazy', isa => Int);

sub _build_number_of_items {
    my $self = shift;
    my @items = $self->items;
    my $total = 0;
    foreach my $i (@items) {
        $total += $i->quantity;
    }
    return $total;
}

sub email {
    return shift->billing_address->email;
}

sub first_name {
    return shift->billing_address->first_name;
}

sub last_name {
    return shift->billing_address->last_name;
}

sub comments {
    return shift->order->{comment_client};
}

sub order_date {
    my $self = shift;
    if (my $date = $self->order->{created}) {
        $date =~ s/ /T/; # eh
        return DateTime::Format::ISO8601->parse_datetime($date);
    }
    return;

}

sub shipping_method {
    return;
}

sub shipping_cost {
    return shift->order->{shipping} || 0;
}

sub subtotal {
    my $self = shift;
    # coupons are not handled yet
    my $subtotal = $self->total_cost - $self->shipping_cost;
    return sprintf('%.2f', $subtotal);
}

sub total_cost {
    return shift->order->{total} || 0;
}

=head2 payment_method

Mapping:


  PP 	= 	Vorauskasse
  CC 	= 	Kreditkarte
  ELV 	= 	Lastschrift
  ELV-AT 	= 	Lastschrift Österreich
  SUE 	= 	Sofortüberweisung
  CB 	= 	ClickAndBuy
  INV 	= 	Rechnung
  INV-AT 	= 	Rechnung Österreich
  PAL 	= 	Paypal
  GP 	= 	giropay
  KLA 	= 	Klarna
  MPA 	= 	mpass
  BAR 	= 	Barzahlen
  YAP 	= 	YAPITAL

=cut

sub payment_method {
    return shift->order->{payment};
}


1;


lib/Marketplace/Rakuten/Order/Item.pm
package Marketplace::Rakuten::Order::Item;

# items must implement sku quantity price subtotal
# remote_shop_order_item
# and a rw merchant_order_item

use strict;
use warnings;

use Moo;
use MooX::Types::MooseLike::Base qw(HashRef Str);
use Data::Dumper;
use namespace::clean;

has struct => (is => 'ro', isa => HashRef);
has merchant_order_item => (is => 'rw', isa => Str);
has order_number => (is => 'ro', default => sub { 'unknown' });

sub quantity {
    return shift->struct->{qty} || 0;
}

=head2 remote_shop_order_item

It looks like the API just passes an id with the position:

 item_id 	Integer 	ID der Bestell-Position

To be meaningful, the class will concatenate the order number with
this.


=cut

sub remote_shop_order_item {
    my $self = shift;
    return $self->order_number . '-' . $self->struct->{item_id};
}

sub price {
    return shift->struct->{price};
}

sub subtotal {
    return shift->struct->{price_sum};
}

=head2 rakuten_id

Either the variant id or the product id. You should have store the ids
when uploading them, so you can use this id to look the real sku up
and store it in the accessor.

=head2 variant_id

The rakuten variant id

=head2 product_id

The rakuten product id

=head2 sku

Rakuten's product_art_no. This appears to map to the actual variant's sku.

=cut

sub rakuten_id {
    my $self = shift;
    return $self->variant_id || $self->product_id;
}

sub variant_id {
    return shift->struct->{variant_id};
}

sub product_id {
    return shift->struct->{product_id};
}

sub sku {
    return shift->struct->{product_art_no};
}

1;


lib/Marketplace/Rakuten.pm
package Marketplace::Rakuten;

use strict;
use warnings;

use Moo;
use MooX::Types::MooseLike::Base qw(Str);
use HTTP::Tiny;
use Marketplace::Rakuten::Response;
use Marketplace::Rakuten::Order;
use Marketplace::Rakuten::Utils;
use namespace::clean;

=head1 NAME

Marketplace::Rakuten - Interface to http://webservice.rakuten.de/documentation

=head1 VERSION

Version 0.03

=head1 CATEGORIES

The list of categories for the German marketplace can be downloaded from
the L<http://api.rakuten.de/categories/csv/download>.

When adding or editing a product, you can pass the
C<rakuten_category_id> to C<add_product> or C<edit_product>

=cut

our $VERSION = '0.03';


=head1 SYNOPSIS

    use Marketplace::Rakuten;
    my $rakuten = Marketplace::Rakuten->new(key => 'xxxxxxx',
                                            endpoint => 'http://webservice.rakuten.de/merchants/'
                                            );
    my $res = $rakuten->get_key_info;

=head1 ACCESSORS

=head2 key

The API key (required).

=head2 endpoint

The URL of the endpoint. Default to http://webservice.rakuten.de/merchants/

If you need a specific api version, it could be something like http://webservice.rakuten.de/v2.05/merchants

=head2 ua

The user-agent. Built lazily.

=cut

has key => (is => 'ro',
            required => 1,
            isa => Str);

has endpoint => (is => 'ro',
                 default => sub { 'http://webservice.rakuten.de/merchants/' });

has ua => (is => 'lazy');

sub _build_ua {
    return HTTP::Tiny->new;
}


=head1 METHODS

=head2 api_call($method, $call, $data);

Generic method to do any call. The first argument is the HTTP request
method (get or post). The second argument is the path of the api, e.g.
(C<misc/getKeyInfo>). The third is the structure to send (an hashref).

Return a Marketplace::Rakuten::Response object.

The key is injected into the data hashref in any case.

=head2 API CALLS

=head3 Miscellaneous

=over 4

=item  get_key_info

=back

=head3 Product creation

=over 4

=item add_product(\%data)

L<http://webservice.rakuten.de/documentation/method/add_product>

Mandatory params: C<name>, C<price>, C<description>

Recommended: C<product_art_no> (the sku)

Return hashref: C<product_id>

=item add_product_image(\%data)

L<http://webservice.rakuten.de/documentation/method/add_product_image>

Mandatory params: C<product_art_no> (the sku) or C<product_id>
(rakuten id).

Return hashref: C<image_id>

=item add_product_variant(\%data)

L<http://webservice.rakuten.de/documentation/method/add_product_variant>

Mandatory params: C<product_art_no> (the sku) or C<product_id>,
C<price>, C<color>, C<label>

Return hashref: C<variant_id>

=item add_product_variant_definition

L<http://webservice.rakuten.de/documentation/method/add_product_variant>

Mandatory params: C<product_art_no> (the sku) or C<product_id>

Recommended: C<variant_1> Size C<variant_2> Color

You need to call this method before trying to upload variants. The
data returned are however just a boolean (kind of) success.

=item add_product_multi_variant

L<http://webservice.rakuten.de/documentation/method/add_product_multi_variant>

Mandatory params: C<product_art_no> (the sku) or C<product_id>, C<variation1_type>, C<variation1_value>, C<price>

Recommended: C<variant_art_no>

Return hashref: C<variant_id>

=item add_product_link

L<http://webservice.rakuten.de/documentation/method/add_product_link>

Mandatory params: C<product_art_no> (the sku) or C<product_id>,
C<name> (100 chars), C<url> (255 chars)

Return hashref: C<link_id>

=item add_product_attribute

L<http://webservice.rakuten.de/documentation/method/add_product_attribute>

Mandatory params: C<product_art_no> (the sku) or C<product_id>,
C<title> (50 chars), C<value>

Return hashref: C<attribute_id>

=back

=cut

sub api_call {
    my ($self, $method, $call, $data) = @_;
    die "Missing method argument" unless $method;
    die "Missing call name" unless $call;

    # populate the data and inject the key
    $data ||= {};
    $data->{key} = $self->key;


    my $ua = $self->ua;
    my $url = $self->endpoint;
    $url =~ s/\/$//;
    $call =~ s/^\///;
    $url .= '/' . $call;

    my $response;
    $method = lc($method);

    if ($method eq 'get') {
        my $params = $ua->www_form_urlencode($data);
        $url .= '?' . $params;
        $response = $ua->get($url);
    }
    elsif ($method eq 'post') {
        $response = $ua->post_form($url, $data);
    }
    else {
        die "Unsupported method $method";
    }
    die "HTTP::Tiny failed to provide a response!" unless $response;
    return Marketplace::Rakuten::Response->new(%$response);
}

sub get_key_info {
    my ($self, $data) = @_;
    $self->api_call(get  => 'misc/getKeyInfo')
}

sub add_product {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/addProduct', $data);
}

sub add_product_image {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/addProductImage', $data);
}

sub add_product_variant {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/addProductVariant', $data);
}

sub add_product_variant_definition {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/addProductVariantDefinition', $data);
}

sub add_product_multi_variant {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/addProductMultiVariant', $data);
}

sub add_product_link {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/addProductLink', $data);
}

sub add_product_attribute {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/addProductAttribute', $data);
}

# unclear what cross-selling is, so leaving it out.

# addProductToShopCategory is referring to an unknown
# shop_category_id, leaving it out for now.


=head3 Product editing

The following methods require and id passed (sku or id) and work the
same as for product adding, but no argument is mandatory.

They all return just a structure with a boolean success (-1 is
failure, 1 is success).

=over 4

=item edit_product(\%data)

Requires C<product_id> or C<product_art_no>

=item edit_product_variant(\%data);

Requires C<variant_id> or C<variant_art_no>

=item edit_product_variant_definition(\%data)

Requires C<product_id> or C<product_art_no>

=item edit_product_multi_variant(\%data);

Requires C<variant_id> or C<variant_art_no>

=item edit_product_attribute

Requires C<attribute_id>, C<title>, C<value>

=back

=cut

sub edit_product {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/editProduct', $data);
}

sub edit_product_variant {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/editProductVariant', $data);
}

sub edit_product_variant_definition {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/editProductVariantDefinition', $data);
}

sub edit_product_multi_variant {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/editProductMultiVariant', $data);
}

sub edit_product_attribute {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/editProductAttribute', $data);
}

=head3 Product deletion

The following methods require just an id or sku. Return a boolean success.

=over 4

=item delete_product(\%data)

Requires C<product_id> or C<product_art_no>

=item delete_product_variant(\%data);

Requires C<variant_id> or C<variant_art_no>

=item delete_product_image(\%data)

Requires C<image_id> (returned by add_product_image).

=item delete_product_link(\%data);

Requires C<link_id> (returned by add_product_link).

=item delete_product_attribute

Requires C<attribute_id> (returned by add_product_attribute).

=back

=cut

sub delete_product {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/deleteProduct', $data);
}

sub delete_product_variant {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/deleteProductVariant', $data);
}

sub delete_product_image {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/deleteProductImage', $data);
}

sub delete_product_link {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/deleteProductLink', $data);
}

sub delete_product_attribute {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/deleteProductAttribute', $data);
}

=head3 Orders

=over 4

=item get_orders(\%params)

No argument is required, but you probably want to pass something like

 { status => pending }

See L<http://webservice.rakuten.de/documentation/method/get_orders>
for the full list of options.

List of statuses:

=over 4

=item pending

=item editable

This is what you want to get for importing.

=item shipped

=item payout

Order is paid. Unclear when the status switches to this one.

=item cancelled

=back

The response is paginated and from here you get only the raw data.

Use get_parsed_orders to get the full list of objects.

=item get_parsed_orders(\%params)

The hashref with the parameters is the same of C<get_orders> (which
get called). This method takes care of the pagination and return a
list of L<Marketplace::Rakuten::Order> objects. You can access the raw
structures with $object->order.

=item get_pending_orders

Shortcut for $self->get_parsed_orders({ status => 'pending' });

=item get_editable_orders

Shortcut for $self->get_parsed_orders({ status => 'editable' });

=cut


sub get_orders {
    my ($self, $data) = @_;
    $self->api_call(get => 'orders/getOrders', $data);
}

sub get_pending_orders {
    my ($self) = @_;
    return $self->get_parsed_orders({ status => 'pending' });
}

sub get_editable_orders {
    my ($self) = @_;
    return $self->get_parsed_orders({ status => 'editable' });
}


sub get_parsed_orders {
    my ($self, $params) = @_;
    $params ||= {};
    my $res = $self->get_orders($params);
    die "Couldn't retrieve orders: " . Dumper($res->errors)
      unless $res->is_success;
    my $data = $res->data;
    my @orders = $self->_parse_orders($data);
    while ($data->{orders}->{paging}->{pages}    and
           $data->{orders}->{paging}->{page}     and
           $data->{orders}->{paging}->{per_page} and
           $data->{orders}->{paging}->{pages} > $data->{orders}->{paging}->{page}) {
        my $next_page = $data->{orders}->{paging}->{page} + 1;
        my $per_page = $data->{orders}->{paging}->{per_page};
        $res = $self->get_orders({
                                  %$params,
                                  page => $next_page,
                                  per_page => $per_page
                                 });
        $data = $res->data;
        push @orders, $self->_parse_orders($data);
    }
    return @orders;
}

sub _parse_orders {
    my ($self, $struct) = @_;
    my @out;
    if ($struct->{orders}) {
        if (my $orders = $struct->{orders}->{order}) {
            foreach my $order (@$orders) {
                my $struct = { %$order };
                Marketplace::Rakuten::Utils::turn_empty_hashrefs_into_empty_strings($struct);
                push @out, Marketplace::Rakuten::Order->new(order => $struct);
            }
        }
    }
    return @out;
}


=item set_order_shipped(\%params)

Required parameter: order_no (the rakuten's order number).

Accepted parameters:

=over 4 

=item dhl (boolean, true if Rakuten-DHL-Rahmenvertrag is used)

=item carrier

=item tracking_number

=item tracking_url

=back

=cut

sub set_order_shipped {
    my ($self, $data) = @_;
    $self->api_call(post => 'orders/setOrderShipped', $data);
}

=item set_order_cancelled(\%params)

Required paramater: C<order_no> (rakuten's order number)

Optional: C<comment>

=cut

sub set_order_cancelled {
    my ($self, $data) = @_;
    $self->api_call(post => 'orders/setOrderCancelled', $data);
}

=item set_order_returned(\%params)

Required paramater: C<order_no> (rakuten's order number) and C<type>
(C<fully> or C<partly>):

L<http://webservice.rakuten.de/documentation/method/set_order_returned>

=back

=cut


sub set_order_returned {
    my ($self, $data) = @_;
    $self->api_call(post => 'orders/setOrderReturned', $data);
}

=head3 Category management

=over 4

=item add_shop_category(\%params)

Required parameter: C<name>

L<http://webservice.rakuten.de/documentation/method/add_shop_category>

Returned: C<shop_category_id>

=item edit_shop_category(\%params)

Required parameter C<shop_category_id> or C<external_shop_category_id>

L<http://webservice.rakuten.de/documentation/method/edit_shop_category>

=item get_shop_categories

No mandatory parameter.

L<http://webservice.rakuten.de/documentation/method/get_shop_categories>

=item delete_shop_category

Required parameter C<shop_category_id> or C<external_shop_category_id>

L<http://webservice.rakuten.de/documentation/method/delete_shop_category>

=item add_product_to_shop_category

Required parameters: C<shop_category_id> or C<external_shop_category_id>,
  C<product_id> or C<product_art_no> (Rakuten's id or our sku).

L<http://webservice.rakuten.de/documentation/method/add_product_to_shop_category>

=back

=cut

sub add_shop_category {
    my ($self, $data) = @_;
    $self->api_call(post => 'categories/addShopCategory', $data);
}

sub edit_shop_category {
    my ($self, $data) = @_;
    $self->api_call(post => 'categories/editShopCategory', $data);
}

sub get_shop_categories {
    my ($self, $data) = @_;
    $self->api_call(get => 'categories/getShopCategories', $data);
}

sub delete_shop_category {
    my ($self, $data) = @_;
    $self->api_call(post => 'categories/deleteShopCategory', $data);
}

sub add_product_to_shop_category {
    my ($self, $data) = @_;
    $self->api_call(post => 'products/addProductToShopCategory', $data);
}

=head1 AUTHOR

Marco Pessotto, C<< <melmothx at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-marketplace-rakuten at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Marketplace-Rakuten>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Marketplace::Rakuten


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Marketplace-Rakuten>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Marketplace-Rakuten>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Marketplace-Rakuten>

=item * Search CPAN

L<http://search.cpan.org/dist/Marketplace-Rakuten/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2015 Marco Pessotto.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of Marketplace::Rakuten
